Add monitoring doc (#419)
SUSE/doc-caasp: adoc/admin-monitoring-health-checks.adoc

----
curl -i http://localhost:10252/healthz
----
* Remote Check (make sure firewall allows port `10252`)
+
[source,bash]
----
curl -i http://<CONTROL-PLANE-IP/FQDN>:10252/healthz
----

- kube-scheduler:
+
The component that schedules workloads to nodes. The URL is http://<CONTROL-PLANE-IP/FQDN>:10251/healthz
+
* Local Check
+
[source,bash]
----
curl -i http://localhost:10251/healthz
----
* Remote Check (make sure firewall allows port `10251`)
+
[source,bash]
----
curl -i http://<CONTROL-PLANE-IP/FQDN>:10251/healthz
----

.High-Availability Environments
[NOTE]
====
In a HA environment you can monitor `kube-apiserver` on `https://<LOAD-BALANCER-IP/FQDN>:6443/healthz`.

If any master node is running correctly you will receive a valid response.

This does, however, not mean that all master nodes necessarily work correctly.
To ensure that all master nodes work properly, the health checks must be repeated individually for each master node deployed.

This endpoint will return a successful HTTP response if the cluster is operational; otherwise it will fail.
It will for example check that it can access `etcd` too.
This should not be used to infer that the overall cluster health is ideal.

> It will return a a successful response even when only minimal operational cluster health exists.

Author: good :eye:

+
* Local Check
+
[source,bash]
----
curl -i http://localhost:10251/healthz
----
* Remote Check (make sure firewall allows port `10251`)
+
[source,bash]
----
curl -i http://<CONTROL-PLANE-IP/FQDN>:10251/healthz
----

.High-Availability Environments
[NOTE]
====
In a HA environment you can monitor `kube-apiserver` on `https://<LOAD-BALANCER-IP/FQDN>:6443/healthz`.

If any master node is running correctly you will receive a valid response.

This does, however, not mean that all master nodes necessarily work correctly.
To ensure that all master nodes work properly, the health checks must be repeated individually for each master node deployed.

This endpoint will return a successful HTTP response if the cluster is operational; otherwise it will fail.
It will for example check that it can access `etcd` too.
This should not be used to infer that the overall cluster health is ideal.
It will return a successful response even when only minimal operational cluster health exists.

To probe for full cluster health, you must perform individual health checking for all machines individually.
====

=== ETCD Cluster

> The etcd cluster expose a endpoint `/health`. The expected (healthy) HTTP response body is `{"health":"true"}`. The etcd cluster is access by HTTPS only. So we need to have etcd certificates.

Reviewer: ```suggestion
The etcd cluster exposes an endpoint `/health`. The expected (healthy) HTTP response body is `{"health":"true"}`. The etcd cluster is accessed by HTTPS only. So we need to have etcd certificates.
```

* Local Check
+
[source,bash]
----
curl --cacert /etc/kubernetes/pki/etcd/ca.crt --cert /etc/kubernetes/pki/etcd/healthcheck-client.crt --key /etc/kubernetes/pki/etcd/healthcheck-client.key https://localhost:2379/health
----
* Remote Check (make sure firewall allows port `2379`)
+
[source,bash]
----
curl --cacert <etcd-root-ca-cert> --cert <etcd-client-cert> --key <etcd-client-key> https://<CONTROL-PLANE-IP/FQDN>:2379/health
----

== Node Health Checks

> Performs basic check of a node health. This consists of checking kubelet endpoint and CNI (Container Networking Interface) pod state.

Reviewer: ```suggestion
Performs a basic check of a node health. This consists of checking the kubelet endpoint and CNI (Container Networking Interface) pod state.
```

=== kubelet

> Is the kubelet up and working in this node?

Reviewer: ```suggestion
Is the kubelet up and working on this node?
```

The kubelet has two ports exposed on all machines:

* Port https/10250: exposes kubelet services to the entire cluster and is available from all nodes using authentication.
* Port http/10248: is only available on local host.

> You can perform an HTTP request to the endpoint to find out if the kubelet is healthy on that machine. The expected (healthy) HTTP response status code `200`.

Reviewer: ```suggestion
You can perform an HTTP request to the endpoint to find out if the kubelet is healthy on that machine. The expected (healthy) HTTP response status code is `200`.
```

metadata:
  name: monitoring
  namespace: kube-system
secrets:
- name: monitoring-secret-token
---
apiVersion: v1
kind: Secret
metadata:
  name: monitoring-secret-token
  namespace: kube-system
  annotations:
    kubernetes.io/service-account.name: monitoring
type: kubernetes.io/service-account-token
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: monitoring-clusterrole
  namespace: kube-system
rules:
- apiGroups: [""]
  resources:
  - nodes/metrics
  - nodes/proxy
  - pods
  verbs: ["get", "list"]
- nonResourceURLs: ["/metrics", "/healthz", "/healthz/*"]
  verbs: ["get"]
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: monitoring-clusterrole-binding
  namespace: kube-system
roleRef:
  kind: ClusterRole
  name: monitoring-clusterrole
  apiGroup: rbac.authorization.k8s.io
subjects:
- kind: ServiceAccount
  name: monitoring
  namespace: kube-system
----
+
Apply the yaml file
+
[source,bash]
----
kubectl apply -f kubelet.yaml
----
Export the token to an environment variable:
+
[source,bash]
----
TOKEN=$(kubectl -n kube-system get secrets monitoring-secret-token -o jsonpath='{.data.token}' | base64 -d)
----
+

> This token can now be passed in headers in the form: "Authorization: Bearer $TOKEN"

Reviewer: ```suggestion
This token can now be passed with headers in the form: "Authorization: Bearer $TOKEN"
```

Author: Or `This token can now be passed in headers in the form`?

Reviewer 2: I reworded it to: This token can now be passed through the `--header` argument as: "Authorization: Bearer $TOKEN".

Author: 💯

+
Now export important values as environment variables.
+

> ** Enviroment Variables Setup

Reviewer: ```suggestion
** Environment Variables Setup
```

.Health check endpoints vs. functional endpoints
[NOTE]
====
A proper health check is always preferred if designed correctly.

Despite the fact that any endpoint could potentially be used to infer if your application is up, a specific health endpoint in your application is preferred.
Such an endpoint will only respond affirmatively when all your setup code on the server has finished and the application is running in a desired state.
====

The `livenessProbes` and `readinessProbes` share configuration options and probe types.

initialDelaySeconds::
Number of seconds to wait before performing the very first liveness probe.

periodSeconds::
Number of seconds that the kubelet should wait between liveness probes.

successThreshold::
Number of minimum consecutive successes for the probe to be considered successful (Default: 1).

failureThreshold::
Number of times this probe is allowed to fail in order to assume that the service is not responding (Default: 3).

timeoutSeconds::
Number of seconds after which the probe times out (Default: 1).

There are different options for the `livenessProbes` to check:

Command::

> A command executed within a container; a reture code of 0 means success. All other return codes mean failure.

Reviewer: ```suggestion
A command executed within a container; a return code of 0 means success. All other return codes mean failure.
```

TCP::
If a TCP connection can be established is considered success.

HTTP::
Any HTTP response between `200` and `400` indicates success.

=== livenessProbe

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/[livenessProbes] are used to detect running but misbehaving pods/a service that might be running (the process didn't die), but that is not responding as expected.

Probes are executed by each `kubelet` against the pods that define them and that are running in that specific node.

When a `livenessProbe` fails, {kube} will automatically restart the pod and increase the `RESTARTS` count for that pod.

These probes will be executed every `periodSeconds` starting from `initialDelaySeconds`.

=== readinessProbe

> https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-readiness-probes[readinessProbes] are used to wait for processes that take some time to start. Despite the container running, it might be performing some time consuming initializatoin operations. During this time, you don't want {kube} to route traffic to that specific pod; also, you don't want that container to be restarted because it will appear unresponsive.

Reviewer: ```suggestion
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-readiness-probes[readinessProbes] are used to wait for processes that take some time to start. Despite the container running, it might be performing some time consuming initialization operations. During this time, you don't want {kube} to route traffic to that specific pod; also, you don't want that container to be restarted because it will appear as unresponsive.
```
